### ISSUE TITLE
Binary input/output not properly processed

### ISSUE THREAD

**Issue by Guufii** [uber/ludwig#323]
Hi,
when I specify a binary input/output and inspect the processed dictionary in Ludwig, I recognize that only features which explicitly use False/True (or FALSE/TRUE) are properly processed as binary inputs/outputs. If I use different notation e.g. 1/0, or big/small or yes/no they are not properly processed, although they are clearly binary. In this case I would have to manually map all binary values to False/True. 

Here is my example input csv:

Binary_1,Binary_2,Binary_3,Binary_4,Binary_5
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true
0,FALSE,yes,big,false
1,TRUE,no,small,true

Model_dict is:

```
model_definition: {   'combiner': {'fc_layers': None, 'type': 'concat'},
    'input_features': [   {   'name': 'Binary_1',
                              'tied_weights': None,
                              'type': 'binary'},
                          {   'name': 'Binary_2',
                              'tied_weights': None,
                              'type': 'binary'},
                          {   'name': 'Binary_3',
                              'tied_weights': None,
                              'type': 'binary'},
                          {   'name': 'Binary_4',
                              'tied_weights': None,
                              'type': 'binary'}],
    'output_features': [   {   'dependencies': [],
                               'loss': {   'confidence_penalty': 0,
                                           'robust_lambda': 0,
                                           'threshold': 0.5,
                                           'weight': 1},
                               'name': 'Binary_5',
                               'reduce_dependencies': 'sum',
                               'reduce_input': 'sum',
                               'threshold': 0.5,
                               'type': 'binary',
                               'weight': 1}],
    'preprocessing': {   'bag': {   'fill_value': '',
                                    'format': 'space',
                                    'lowercase': 10000,
                                    'missing_value_strategy': 'fill_with_const',
                                    'most_common': False},
                         'binary': {   'fill_value': 0,
                                       'missing_value_strategy': 'fill_with_const'},
                         'category': {   'fill_value': '<UNK>',
                                         'lowercase': False,
                                         'missing_value_strategy': 'fill_with_const',
                                         'most_common': 10000},
                         'force_split': False,
                         'image': {   'in_memory': True,
                                      'missing_value_strategy': 'backfill',
                                      'resize_method': 'crop_or_pad'},
                         'numerical': {   'fill_value': 0,
                                          'missing_value_strategy': 'fill_with_const'},
                         'sequence': {   'fill_value': '',
                                         'format': 'space',
                                         'lowercase': False,
                                         'missing_value_strategy': 'fill_with_const',
                                         'most_common': 20000,
                                         'padding': 'right',
                                         'padding_symbol': '<PAD>',
                                         'sequence_length_limit': 256,
                                         'unknown_symbol': '<UNK>'},
                         'set': {   'fill_value': '',
                                    'format': 'space',
                                    'lowercase': False,
                                    'missing_value_strategy': 'fill_with_const',
                                    'most_common': 10000},
                         'split_probabilities': (0.7, 0.1, 0.2),
                         'stratify': None,
                         'text': {   'char_format': 'characters',
                                     'char_most_common': 70,
                                     'char_sequence_length_limit': 1024,
                                     'fill_value': '',
                                     'lowercase': True,
                                     'missing_value_strategy': 'fill_with_const',
                                     'padding': 'right',
                                     'padding_symbol': '<PAD>',
                                     'unknown_symbol': '<UNK>',
                                     'word_format': 'space_punct',
                                     'word_most_common': 20000,
                                     'word_sequence_length_limit': 256},
                         'timeseries': { 'fill_value': '',
                                           'format': 'space',
                                           'missing_value_strategy': 'fill_with_const',
                                           'padding': 'right',
                                           'padding_value': 0,
                                           'timeseries_length_limit': 256}},
    'training': {   'batch_size': 128,
                    'bucketing_field': None,
                    'decay': False,
                    'decay_rate': 0.95,
                    'decay_steps': 10000,
                    'dropout_rate': 0.05,
                    'early_stop': 7,
                    'epochs': 10,
                    'eval_batch_size': 0,
                    'gradient_clipping': None,
                    'increase_batch_size_on_plateau': 0,
                    'increase_batch_size_on_plateau_max': 512,
                    'increase_batch_size_on_plateau_patience': 5,
                    'increase_batch_size_on_plateau_rate': 2,
                    'learning_rate': 0.001,
                    'learning_rate_warmup_epochs': 5,
                    'optimizer': {   'beta1': 0.9,
                                     'beta2': 0.999,
                                     'epsilon': 1e-08,
                                     'type': 'adam'},
                    'reduce_learning_rate_on_plateau': 0,
                    'reduce_learning_rate_on_plateau_patience': 5,
                    'reduce_learning_rate_on_plateau_rate': 0.5,
                    'regularization_lambda': 0,
                    'regularizer': 'l2',
                    'staircase': False,
                    'validation_field': 'combined',
                    'validation_measure': 'loss'}}
```

The output should be alternating binary values. However, if I inspect dict of the preprocessed data set in '_predict' line 681, this is only the case feature Binary_2 and Binary_5. 

Hope you are able to reproduce the bug. 

BTW. Thanks for this very useful piece of software!

Best
G

**Comment by w4nderlust** [comment]
You are welcome @guufi !
I tested running the same code that is used for preprocessing binary features (`features/binary/features.py:61`) on the csv you provided. Column 1, 2 and 5 work, while 3 and 4 are mapped to all `True`. To be honest, I believe that this is pretty ok, if you have arbitrary strings like big and small you can always use a categorical feature instead and specify the representation to be sparse (meaning 1 hot encoding) and btain the same thing. Also if you really really want to use a binary feature, then you can manipulate your data to be either 0 and 1 or true and false.

**Comment by Guufii** [comment]
Hi @w4nderlust ,
thank you for the hint with the categorical variable I will try that as an alternative. However, even the input of 0/1 is not recognized as binary. See screenshot with input and output to the function you mentioned in the debugger. The "pass" at the end is only to have a line to stop with a break point. I use numpy 1.16.3 and pandas 0.24.2. In my case the alternating zeros and ones are mapped all to true.
![image](https://user-images.githubusercontent.com/49252109/57215913-a43ef600-6fee-11e9-9c54-d9c0ebceb638.png)

Hope you can reproduce this behavior. 

Best
G

**Comment by Guufii** [comment]
How about replacing line 61, with:

unique_list = dataset_df[feature['name']].unique()
        try:
            data[feature['name']] = dataset_df[feature['name']].map(
            {unique_list[0]: False, unique_list[1]: True}).as_matrix()
        except ValueError as e:
            raise ValueError(
                'Binary features need to have exactly two different values!'
            )

Best G

**Comment by w4nderlust** [comment]
This is doable, but it entails another problem: at test time, datapoints will have those string values too, so the mapping from string to 0 and 1 will need to be saved somewhere, specifically in the metadata file, while now nothing from binary features is saved in the metadata file.
I see the advantage in doing it, but at the same time I'm not sure of how big of a deal this is. What do you think?

**Comment by Guufii** [comment]
Hi @w4nderlust,
I am not sure about the implication of this modification in the metadata file. But I see point your point about the mapping at test time. From the inspection of the source I would assume that most of the functionality can be taken from category features. So actual coding effort should be minimal. But you are the expert and have a deeper knowledge about all caveats this might bring.

Best
G

**Comment by w4nderlust** [comment]
Yeah coding is not the issue, it would be a simple change, it's more a matter of opportunity.
For now I will postpone it, in particular because there are many more urgent tasks to do, but I may reconsider it later.
Thank really much anyway for looking into this thing and suggesting solutions, it is much appreciated!
